website_multi_company/models/website.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
+
+
+class Website(models.Model):
+    _inherit = 'website'
+
+    @api.multi
+    @api.constrains('user_id')
+    def _check_user_id(self):
+        """Do not allow same public user on two websites.s"""
+        for record in self:
+            existing = self.search([('user_id', '=', record.user_id.id)])
+            if len(existing) > 1:
+                raise ValidationError(_(
+                    'You cannot assign the same public user to two websites.',
+                ))
+
+    @api.multi
+    @api.constrains('user_id.company_id', 'company_id')
+    def _check_user_id_company_id(self):
+        """Do not allow differing company on public user and website."""
+        for record in self:
+            if record.company_id != record.user_id.company_id:
+                raise ValidationError(_(
+                    'The public user company must match the website company.',
+                ))
+
+    @api.model
+    def create(self, vals):
+        """Assign a unique public user per website and align the company."""
+        if not vals.get('user_id'):
+            user = self._get_new_public_user()
+            company_id = vals.get('company_id') or self.env.user.company_id.id
+            if user.company_id.id != company_id:
+                user.company_id = company_id
+            vals['user_id'] = user.id
+        user = super(Website, self).create(vals)
+        # This is required to circumvent duplicate keys
+        user.login = 'public-%d' % user.id
```

Comment:
What about user's name?

Reply:
This is for the public user login itself. The user defined login is used when the actual user is created.

Without this fix, we can only have two websites. The third website (and subsequent third public user) will yield a duplicate key error on login for `public (copy)`

Reply:
I need to drink coffee before things. Ok so yeah `name` doesn't have a unique constraint, but I can see how the canonical would be good here. Updating.